bin/gaia_marionette_retry.js

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env node
+var Promise = require('es6-promise').Promise;
+var format = require('util').format;
+var path = require('path');
+var spawn = require('child_process').spawn;
+
+function main() {
+  // 0 => node
+  // 1 => ./bin/gaia_marionette_retry.js
+  process.argv.splice(0, 2);
+
+  var index = 0;
+  while (process.argv[index].slice(0, 2) === '--') {
+    index += 2;
+  }
+
+  var args = process.argv.splice(0, index);
+  var filenames = process.argv;
+
+  // Also look at the end of the cli invocation for cli args.
+  var count = 0;
+  index = filenames.length;
+  while (filenames[index - 2].slice(0, 2) === '--') {
+    count += 2;
+    index -= 2;
+  }
+
+  args = args.concat(filenames.splice(index, count));
+
+  if (filenames[0].indexOf('setup') !== -1) {
+    // Special case for setup script.
+    args.push(filenames.shift());
+  }
+
+  runTests(filenames, args, 5 /* retry */).then(function(results) {
+    // Summarize.
+    console.log('*~*~* Results *~*~*');
+    console.log('passed: %d', results.pass);
+    console.log('failed: %d', results.fail);
+    console.log('todo: %d', results.pending);
+  });
+}
+
+function runTests(filenames, args, retry) {
+  if (filenames.length === 0) {
+    // All done!
+    return Promise.resolve({ pass: 0, fail: 0, pending: 0 });
+  }
+
+  var next = filenames.pop();
+  var tally;
+  return runTests(filenames, args, retry)
+  .then(function(results) {
+    tally = results;
+    return runTest(next, args, retry);
+  })
+  .then(function(result) {
+    // Now we have to output our result and package it
+    // so that it can be aggregated for final results.
+    var stdout = result.stdout;
+    var incremental = stdout.slice(0, stdout.indexOf('*~*~*'));
+    // Print incremental result.
+    process.stdout.write(incremental);
+    var pass = parseInt(stdout.match(/passed:\s*(\d+)/)[1]);
+    var fail = parseInt(stdout.match(/failed:\s*(\d+)/)[1]);
+    var pending = parseInt(stdout.match(/todo:\s*(\d+)/)[1]);
+    tally.pass += pass;
+    tally.fail += fail;
+    tally.pending += pending;
+    return Promise.resolve(tally);
+  })
+  .catch(function(error) {
+    console.error('Something is really wrong!!');
+    console.error(error.toString());
+  });
+}
+
+function runTest(filename, args, retry) {
```

Comment:
Good comments. I wonder if we should copy-paste them into the file before the functions? :)